cmd/pi-strings-go/strings.go

```diff
@@ -0,0 +1,191 @@
+
+import (
+	"sort"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	stringspb "github.com/lapsang-boys/pippi/proto/strings"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/traditionalchinese"
+	textunicode "golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/encoding/unicode/utf32"
+	"golang.org/x/text/transform"
+)
+
+// extractStrings extracts printable strings with a minimum number of characters
+// from the given binary.
+func extractStrings(buf []byte, minLength int) []*stringspb.StringInfo {
+	c := make(chan []*stringspb.StringInfo)
+	encs := []struct {
+		enc stringspb.Encoding
+		dec *encoding.Decoder
+	}{
+		// * UTF-8
+		{
+			enc: stringspb.Encoding_UTF8,
+			dec: textunicode.UTF8.NewDecoder(),
+		},
+		// * UTF-16
+		//    - big endian
+		{
+			enc: stringspb.Encoding_UTF16BigEndian,
+			dec: textunicode.UTF16(textunicode.BigEndian, textunicode.IgnoreBOM).NewDecoder(),
+		},
+		//    - big endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF16BigEndianBOM,
+			dec: textunicode.UTF16(textunicode.BigEndian, textunicode.ExpectBOM).NewDecoder(),
+		},
+		//    - little endian
+		{
+			enc: stringspb.Encoding_UTF16LittleEndian,
+			dec: textunicode.UTF16(textunicode.LittleEndian, textunicode.IgnoreBOM).NewDecoder(),
+		},
+		//    - little endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF16LittleEndianBOM,
+			dec: textunicode.UTF16(textunicode.LittleEndian, textunicode.ExpectBOM).NewDecoder(),
+		},
+		// * UTF-32
+		//    - big endian
+		{
+			enc: stringspb.Encoding_UTF32BigEndian,
+			dec: utf32.UTF32(utf32.BigEndian, utf32.IgnoreBOM).NewDecoder(),
+		},
+		//    - big endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF32BigEndianBOM,
+			dec: utf32.UTF32(utf32.BigEndian, utf32.ExpectBOM).NewDecoder(),
+		},
+		//    - little endian
+		{
+			enc: stringspb.Encoding_UTF32LittleEndian,
+			dec: utf32.UTF32(utf32.LittleEndian, utf32.IgnoreBOM).NewDecoder(),
+		},
+		//    - little endian, with BOM
+		{
+			enc: stringspb.Encoding_UTF32LittleEndianBOM,
+			dec: utf32.UTF32(utf32.LittleEndian, utf32.ExpectBOM).NewDecoder(),
+		},
+		// * Big-5
+		{
+			enc: stringspb.Encoding_Big5,
+			dec: traditionalchinese.Big5.NewDecoder(),
+		},
+		// TODO: consider adding support for ISO-88xx-xx encodings (e.g. 8859-1
+		// for latin)?
+		//
+		//    * https://godoc.org/golang.org/x/text/encoding/charmap
+
+		// TODO: add support for more encodings? got to catch 'em all!
+		//
+		//    * https://godoc.org/golang.org/x/text/encoding/japanese
+		//    * https://godoc.org/golang.org/x/text/encoding/korean
+		//    * https://godoc.org/golang.org/x/text/encoding/simplifiedchinese
+
+		// TODO: add NULL-terminated string encoding.
+
+		// TODO: add length-prefixed string encoding.
+	}
+	for _, enc := range encs {
+		go extractEncStrings(buf, minLength, enc.enc, enc.dec, c)
+	}
+	var infos []*stringspb.StringInfo
+	for range encs {
+		infos = append(infos, <-c...)
+	}
+	// Sort results.
+	sort.Slice(infos, func(i, j int) bool {
```

Comment:
Is it an idiom to inline less function inside sort? Otherwise I'd argue that it's worth to save the mental overload by creating a `less` function.

Reply:
I'm fine with having a `less` function. Using function literals for this seems quite idiomatic.

From Go 1.13 source tree:
```go
src/flag/flag.go:	sort.Slice(result, func(i, j int) bool {
src/compress/bzip2/huffman.go:	sort.Slice(pairs, func(i, j int) bool {
src/compress/bzip2/huffman.go:	sort.Slice(codes, func(i, j int) bool {
src/time/genzabbrs.go:	sort.Slice(zs, func(i, j int) bool {
src/go/doc/example.go:	sort.Slice(list, func(i, j int) bool {
src/go/doc/example.go:	sort.Slice(decls, func(i, j int) bool {
src/go/doc/example.go:	sort.Slice(comments, func(i, j int) bool {
src/go/ast/import.go:	sort.Slice(specs, func(i, j int) bool {
src/go/ast/import.go:	sort.Slice(comments, func(i, j int) bool {
src/internal/trace/goroutines.go:		sort.Slice(g.Regions, func(i, j int) bool {
src/internal/trace/mud.go:	sort.Slice(edges, func(i, j int) bool {
src/encoding/json/encode.go:	sort.Slice(sv, func(i, j int) bool { return sv[i].s < sv[j].s })
src/encoding/json/encode.go:	sort.Slice(fields, func(i, j int) bool {
src/cmd/go/internal/module/module.go:	sort.Slice(list, func(i, j int) bool {
src/cmd/go/internal/modload/import.go:		sort.Slice(mods, func(i, j int) bool {
src/net/http/fs.go:	sort.Slice(dirs, func(i, j int) bool { return dirs[i].Name() < dirs[j].Name() })
src/net/http/pprof/pprof.go:	sort.Slice(profiles, func(i, j int) bool {
src/net/http/cookiejar/jar.go:	sort.Slice(selected, func(i, j int) bool {
src/runtime/pprof/pprof.go:	sort.Slice(all, func(i, j int) bool { return all[i].name < all[j].name })
src/runtime/pprof/pprof.go:	sort.Slice(all, func(i, j int) bool {
src/runtime/pprof/pprof.go:	sort.Slice(p, func(i, j int) bool { return p[i].InUseBytes() > p[j].InUseBytes() })
src/runtime/pprof/pprof.go:	sort.Slice(p, func(i, j int) bool { return p[i].Cycles > p[j].Cycles })
src/runtime/pprof/pprof.go:	sort.Slice(p, func(i, j int) bool { return p[i].Cycles > p[j].Cycles })
src/runtime/debug/garbage.go:			sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
src/cmd/go/internal/mvs/mvs.go:	sort.Slice(tail, func(i, j int) bool {
src/cmd/go/internal/mvs/mvs.go:	sort.Slice(min, func(i, j int) bool {
src/cmd/go/internal/modcmd/graph.go:	sort.Slice(out[deps:], func(i, j int) bool {
src/cmd/go/internal/modfile/rule.go:		sort.Slice(block.Line, func(i, j int) bool {
src/cmd/go/internal/modfetch/repo.go:	sort.Slice(list, func(i, j int) bool {
src/cmd/go/internal/envcmd/env.go:	sort.Slice(lines, func(i, j int) bool {
src/cmd/internal/obj/objfile.go:		sort.Slice(ctxt.Data, func(i, j int) bool {
src/cmd/compile/internal/gc/main.go:	sort.Slice(largeStackFrames, func(i, j int) bool {
src/cmd/compile/internal/gc/main.go:	sort.Slice(unused, func(i, j int) bool { return unused[i].pos.Before(unused[j].pos) })
src/cmd/compile/internal/gc/syntax.go:	sort.Slice(res, func(i, j int) bool { return less(res[i], res[j]) })
src/cmd/compile/internal/gc/iexport.go:		sort.Slice(objs, func(i, j int) bool {
src/cmd/compile/internal/gc/iexport.go:	sort.Slice(pkgs, func(i, j int) bool {
src/cmd/compile/internal/gc/obj.go:	sort.Slice(funcsyms, func(i, j int) bool {
src/cmd/compile/internal/gc/pgen.go:			sort.Slice(compilequeue, func(i, j int) bool {
src/cmd/trace/goroutines.go:	sort.Slice(glist, func(i, j int) bool { return glist[i].ExecTime > glist[j].ExecTime })
src/cmd/trace/goroutines.go:	sort.Slice(glist, func(i, j int) bool {
src/cmd/trace/annotations.go:	sort.Slice(userTasks, func(i, j int) bool {
src/cmd/trace/annotations.go:	sort.Slice(userRegions, func(i, j int) bool {
src/cmd/trace/annotations.go:	sort.Slice(data, func(i, j int) bool {
src/cmd/trace/annotations.go:		sort.SliceStable(rawEvents, func(i, j int) bool { return rawEvents[i].Ts < rawEvents[j].Ts })
src/cmd/trace/annotations.go:	sort.Slice(data, func(i, j int) bool {
src/cmd/trace/annotations.go:		sort.SliceStable(task.regions, func(i, j int) bool {
src/cmd/trace/pprof.go:		sort.Slice(intervals, func(i, j int) bool {
src/cmd/trace/trace.go:		sort.SliceStable(sortedTask, func(i, j int) bool {
src/cmd/vendor/golang.org/x/tools/go/analysis/internal/analysisflags/help.go:		sort.Slice(analyzers, func(i, j int) bool {
src/cmd/vendor/golang.org/x/tools/go/analysis/internal/facts/facts.go:	sort.Slice(gobFacts, func(i, j int) bool {
src/go/types/methodset.go:	sort.Slice(list, func(i, j int) bool {
src/go/types/stmt.go:	sort.Slice(unused, func(i, j int) bool {
src/io/ioutil/ioutil.go:	sort.Slice(list, func(i, j int) bool { return list[i].Name() < list[j].Name() })
src/cmd/nm/nm.go:			sort.Slice(syms, func(i, j int) bool { return syms[i].Addr < syms[j].Addr })
src/cmd/nm/nm.go:			sort.Slice(syms, func(i, j int) bool { return syms[i].Name < syms[j].Name })
src/cmd/nm/nm.go:			sort.Slice(syms, func(i, j int) bool { return syms[i].Size > syms[j].Size })
src/cmd/go/internal/modget/get.go:		sort.Slice(lostUpgrades, func(i, j int) bool {
```

Reply:
Than, it's fine! I suspected it could be idiomatic